### ISSUE TITLE
Allow package authors to run all NuGet.org validations locally through the NuGet/DotNet CLI.

### ISSUE THREAD

**Issue by aortiz-msft** [NuGet/Home#12575]
### NuGet Product(s) Involved

dotnet.exe

### The Elevator Pitch

Authors want to be able to run validations on their packages before uploading to NuGet.org.

**Comment by agr** [comment]
What's the expected utility? Save on bandwidth consumed trying to upload the package?

There are two kinds of validations nuget.org does: "sync" and "async". "Sync" validation failure would result in "4xx" HTTP response at push time, i.e. immediate feedback (at the cost of round trip to server). "Async" validations involve private external (to the NuGet server team) services, validations from which we won't be able to ship with a client.

A lot of "async" validation failures are a result of a race condition with package author signing: the very last step of "async" validation is verifying that package signatures (both author and repository) are good, so if a package author submits an unsigned package, then immediately would go to account settings and add author signature, that last step will fail as the package wouldn't have author signature while owner account would require one. That's not something that we can prevent from the client side. We can check for presence of author signature, but we need to know if publishing would require one, which in turn mean a round trip to server, so why not just try uploading the package?

I.e. if the goal is to reduce package pushes that result in 4xx responses, we can have most of validations that cause those client side (they'd have to be optional in case there is no intent to publish to NuGet.org), if the goal is to reduce the amount of async failures, I'm afraid it is not as simple.

**Comment by agr** [comment]
https://github.com/NuGet/Home/issues/12566 - related?

**Comment by JonDouglas** [comment]
@agr I think this would be considered a similar experience as to `--dry-run` in many other package push/publish scenarios across other ecosystems. In other words, could someone have confidence to say that the dry run was successful and the only failure points might be said private / external services validations that are unique to the registry itself, but not the package. 

In other words, could we decouple package validations and registry validations to provide authors "package validations" locally before going through registry validations.

**Comment by erdembayar** [comment]
> 
> In other words, could we decouple package validations and registry validations to provide authors "package validations" locally before going through registry validations.

Don't we show warning message about package during `pack operation`? `Pack operation` does several validations and show messages. Should we extend that one or reuse some logics from it?

**Comment by zivkan** [comment]
> What's the expected utility?

[I proposed this in another issue](https://github.com/NuGet/Home/issues/12566#issuecomment-1527170836) where a customer created packages which failed validation when pushing to nuget.org.  In other words, customers want the maximum confidence possible that packages will be accepted on a feed at pack time, so can be validated before pushing (for example, when creating pull requests).

>  I think this would be considered a similar experience as to --dry-run in many other package push/publish scenarios

That's not what I intended when I commented in the other issue. A `--dry-run` is a good idea, but not what I believe this issue is about.

> the only failure points might be said private / external services validations that are unique to the registry itself

This is exactly what I believe this feature should enable/validate.

> if the goal is to reduce the amount of async failures, I'm afraid it is not as simple.

Perfect shouldn't be the enemy of good enough. There are other feed-specific validations that absolutely can be validated client side at pack time. Examples I think about:

* File size limitations
* Any static file validations (for example, files not allowed to be in snupkg)
* Are nuspec files allowed to contain `<owner>` element?
* The [nuspec reference docs](https://learn.microsoft.com/en-us/nuget/reference/nuspec) has the text "when uploading a package to nuget.org" 14 times.

I also don't believe this should be nuget.org specific. We should design this to be extensible via plugins, so that

1. it doesn't require changing NuGet and shipping a new version of nuget.exe or the dotnet CLI to customers to get updates if nuget.org changes server-side validations.
2. Any feed can implement their own validations.
   * I know that Azure Artifacts has a different file size limit, and I don't believe they support either `.symbols.nupkg` or `.snupkg` packages.
   * GitHub Package Repository requires packages to contain a repo URL in the nuspec, that starts with `https://github.com/` (in fact, I think  the URL must be in the same org/account as where the package is being pushed to, but pack static validation isn't going to know where it's going to be pushed, so a github URL prefix validation might be good enough to point customers starting out in the right direction).
